Slices rebased again

djdiskmachine opened the pull request:
Not quite sure what to do with this section that went in MANUAL.md?

  - slicer will cut the sample into "slices" (displayed only in slice mode) amount of samples. Slices are mapped from C-2 (the lowest note) up to amount of slices. Example: slices == 4 will give you four slices mapped to C-2, C#-2, D-2, D#-2

Commit:
Adding slices

Rebased on master
Issue when loading project "restoring allocation" whenever 164-170 are added

Commit:
Manually fix clang errors

Commit:
Dynamically show slices

Show or hide slices field by refreshing instrument screen when changing loop mode

Commit:
Slices aren't pointers anymore

Commit:
Fix legacy errors

maks commented:
Awesome, thanks so much for working on this again @djdiskmachine ! 👍🏻 

> Not quite sure what to do with this section that went in MANUAL.md?
We have the proper picoTraacker user manual now so I would put it in `usermanual/content/pages/instruments.md`

djdiskmachine committed:
Show / hide slices to the right of loopmode

  Dynamically show / hide slices field
  Slices field to the right of loop mode to respect display layout
  Redraws using onInstrumentChange
  Adds change listener in InstrumentView::Update
  Creates setFocus function to save cursor placement

djdiskmachine commented:
Tested Nd working on Pico tracker v1
![PXL_20250415_061822829.MP.jpg](https://github.com/user-attachments/assets/2bddb42b-fdfa-4dab-95b7-754dfbfccaab)

![PXL_20250415_061653853.MP.jpg](https://github.com/user-attachments/assets/c3b85575-9c56-4c69-aa71-0903f8a3c67f)

democloid commented on sources/Application/Instruments/SampleInstrument.cpp:
```diff
@@ -88,6 +89,7 @@ SampleInstrument::SampleInstrument()
   loopEnd_.AddObserver(*this);
   variables_.insert(variables_.end(), &table_);
   variables_.insert(variables_.end(), &tableAuto_);
+  variables_.insert(variables_.end(), &slices_);
```
We need to increase the number of slots on the `variables_` list

democloid commented on sources/Application/Views/InstrumentView.cpp:
```diff
@@ -369,6 +364,14 @@ void InstrumentView::fillSampleParameters() {
+    position._x += 18;
+    v = instrument->FindVariable(FourCC::SampleInstrumentSlices);
+    intVarField_.emplace_back(position, *v, "slcs: %d", 1, 128, 1, 16);
+    fieldList_.insert(fieldList_.end(), &(*intVarField_.rbegin()));
```
Need to increase the `fieldList_` list size

democloid reviewed (changes requested):
Haven't tested this yet, couple of things we need to fix.

On a general note, is the slicer a loop mode? or should it be a separate property of the sample which we could then apply a loop mode to it on a per slice basis?

Commit:
Update list sizes

djdiskmachine commented:
> On a general note, is the slicer a loop mode? or should it be a separate property of the sample which we could then apply a loop mode to it on a per slice basis?

If you'd prefer this behavior I'll happily accept a design proposal. :)

maks commented on sources/Foundation/Types/Types.h:
```diff
@@ -56,6 +56,7 @@ struct FourCC {
     SampleInstrumentEnd = 6,
     SampleInstrumentTable = 117,
     SampleInstrumentTableAutomation = 60,
+    SampleInstrumentSlices = 202,
```
Please use the next available number and add this usage at line 162. I think this is my fault as I incorrectly allocated a couple of new consts as 200, 201 but I have since fixed that in `master` so the next available number should be `151.

maks reviewed (changes requested):
I still need to test this myself but all looks good except a few small nitpicks.

djdiskmachine commented on sources/Application/Instruments/SampleInstrumentDatas.h:
```diff
@@ -2,7 +2,7 @@
 const char *loopTypes[SILM_LAST] = {"none", "loop", "pingpong", "oscillator",
                                     //	"oscillator fine",
-                                    "looper sync"};
+                                    "looper sync", "slicer"};
```
I don't think so, slicer is the mode, slices is the count. :)

maks replied:
Well from the users point of view the loop _mode_ is "slices" of a sample, slicer is the action or actor involved. 
This would also let the following field be labeled "num" it would read a bit more more naturally:
```
loop mode: slices  num: 16
```

djdiskmachine committed:
Fix review comments
Change to next available number in Types.h
Rename loop mode to slices and modify string comparison in InstrumentView.cpp
Wrap single line conditional in gull wings

maks reviewed (approved):
thanks again @djdiskmachine ! 👍🏻

democloid commented:
I'll close this one in favor of #504
